### ISSUE TITLE
Kernel source & config

### ISSUE THREAD

**Issue by ankurkotwal** [thefossguy/archlinux-visionfive2#2]
Hey @thefossguy this image is great!
How did you build the kernel? Yours doesn't support the USB wifi dongle that was provided by Starfive but their image v2.10.4 does. I'd therefore like to build a version of the kernel that does support it.

**Comment by thefossguy** [comment]
Hi Ankur, thank you for the feedback! I really appreciate it.

I built the kernel using a custom PKGBUILD. ~~[Here is the repo](https://git.thefossguy.com/thefossguy/linux-starfive-visionfive2). I've yet to mirror it on GitHub but you should find it under this account with the same repo name soon!~~

About the USB WiFi dongle, I did not order it at the time so don't have physical hardware to test if it is working, with or without any issues. So I haven't yet bothered to look into it. Strictly speaking, I don't need the wireless dongle, but if I can obtain it, I'd be happy to enable the support for it.

Since you mention that it is working in the vendor Debian image, can you send me the config with which the vendor kernel was built? You can obtain it with either of the following commands:

```bash
$ zcat /proc/config.gz > ~/.config
$ cp /boot/config-$(uname -r) > ~/.config
```

Once that is done, I will also need to know what modules are loaded, so I can pin point which config option enables the USB dongle support. You can get that using the `lsmod` command, like so:

```bash
$ lsmod > ~/loaded_modules
```

Once that is done, you can attach the files `.config` and `loaded_modules` in this issue thread from your home directory and remove them once they are uploaded.

**Edit**: The kernel package repository is now mirrored on GitHub and is now [available here](https://github.com/thefossguy/linux-starfive-visionfive2).

---

P.S. I'd love to know how you came accross this image. Since I delibrately didn't install any display output related packages--to keep the image as small as possible, yet functional for porting purposes--I haven't advertised it anywhere but in some IRC channels. Just curious ;)

**Comment by ankurkotwal** [comment]
Thanks for the super fast response.
Here's the config starfive uses: 
[starfive_debian_202302.tar.gz](https://github.com/thefossguy/archlinux-visionfive2/files/10887602/starfive_debian_202302.tar.gz)

In terms of how I found your repo, a simple "arch linux risc-v" Google search brought me to https://wiki.archlinux.org/title/User:Cmsigler/RISC-V which refers to an Arch image builder at https://github.com/sehraf/d1-riscv-arch-image-builder That repo refers to this one.

For the last few weeks, I've been trying to build my own Arch image using Starfive's kernel. I've been in touch with cwt who published the first image on rvspace as well. The issue I have been having is that with any of starfive's kernels, the Arch tty won't accept my password. It accepts my username but as soon as I type a single character into the password field, it submits it and fails. This issue does not occur on your  image nor cwt's image. I'm keen on building my own arch image so your repo is perfect for what I need, except I couldn't see how you were getting the kernel.

BTW I've found a few issues with your scripts that I'll submit PRs for in the coming days. A summary of the issues found (I'm using a fresh Manjaro deskop install for this work):
- dash isn't installed and your scripts assume it. I propose adopting bash instead as its available everywhere.
- No check for running as root so lots of things will fail, the first being losetup
- Timezone is India, perhaps grab from host machine.
- Hostname presumably has author’s name. Consider something generic.
- Doesn't check for pacstrap being on the system (package name arch-install-scripts).
- Build is currently failing with "Partition /mnt too full: 484252 blocks needed, 383253 blocks free". Did your package install also fetch 233 packages? Seems like a lot for a minimal image.
- Duplication of some packages such as vim and neovim. I propose having a minimal list but allowing the user to add a file with a list of additional packages they might want. That way you could separate the completely minimal image from something folks want to extend.
- Consider having a different root password like "starfive". That way for something connected to the public internet, folks won't easily guess the password.
- Consider disallowing root login over ssh by default and forcing login via the user (sudo).

Thanks for all your work on this so far. I'm looking forward to getting it going!

**Comment by thefossguy** [comment]
Thank you for providing the requested files! I'll look into enabling support for the USB dongle. I will soon reach out to you with a kernel binary that hopefully supports the USB dongle.

The issues that you list out are sound and what I would expect how a third party image might be. So thank you for the feedback! But bear in mind, this is still very rough work. I have yet to look into hanling some of the issues. Like the timezone, I used it because this image was primarily for myself but I'm looking to make it generic for everyone by using defaults like UTC for timezone, removing a few "personal" packages like `bat`, etc.

If you solve listed issues before me, I'd be happy to accept your PR and maintain the image! :)

**Comment by ankurkotwal** [comment]
> Thank you for providing the requested files! I'll look into enabling support for the USB dongle. I will soon reach out to you with a kernel binary that hopefully supports the USB dongle.

Thank you!

> The issues that you list out are sound and what I would expect how a third party image might be. So thank you for the feedback! But bear in mind, this is still very rough work.

Oh, sorry if I made it seem like I expect this to be fixed. I'm happy to help contribute PRs.

> I have yet to look into hanling some of the issues. Like the timezone, I used it because this image was primarily for myself but I'm looking to make it generic for everyone by using defaults like UTC for timezone, removing a few "personal" packages like `bat`, etc.

I'm also a `bat` fan! Allowing the user to provide a conf file with extra packages, hostname, default user, password, etc might make this a lot easier to customise. I can look into that.

Any ideas about the size issue? I'm not familiar enough with pacstrap to know where it's fetching the packages from. Since I'm building on an x86 is it defaulting to x86 packages? Which repo is it fetching from? It should be fetching from the repos listed at https://archriscv.felixc.at/ right?

> 
> If you solve listed issues before me, I'd be happy to accept your PR and maintain the image! :)
You will see some coming your way :)

**Comment by thefossguy** [comment]
>> The issues that you list out are sound and what I would expect how a third party image might be. So thank you for the feedback! But bear in mind, this is still very rough work.
>
> Oh, sorry if I made it seem like I expect this to be fixed. I'm happy to help contribute PRs.

Oh no, quite the opposite. Was letting you know the current state of things :)

> Any ideas about the size issue?

That might be because I am using a few meta packages like `base` and `base-devel` which pull a lot of packges. The former is necessary for the base system and the later is installed because this image is for developers (who are porting their software) rather than end users :)

> I'm not familiar enough with pacstrap to know where it's fetching the packages from. Since I'm building on an x86 is it defaulting to x86 packages?

Maybe, I can't say for sure without looking at your `pacman.conf` file but I'd guess that if you are on an x86 machine, you'd be pulling in x86 packages. Specifying only the architecture won't help much since the official Arch Linux repositories don't have non-x86 packages.

> Which repo is it fetching from?

Initially, I stared with cwt's image, which was already using https://archriscv.felixc.at/ as the repository, so that was handled for me. (Thanks cwt!) And this image was built natively on the VF2 to there was no architecture mismatch for packages.

(Apologies for the late reply, I just saw your PR and got reminded that I forgot to reply to this.)

**Comment by thefossguy** [comment]
> Thanks for the super fast response.
> Here's the config starfive uses:
> [starfive_debian_202302.tar.gz](https://github.com/thefossguy/archlinux-visionfive2/files/10887602/starfive_debian_202302.tar.gz)

I assume that the output of `lsmod` was provided form the Debian image (I don't recall mentioning this earlier, apologies). In the provided output, I see the module `wlan_ecr6600u_usb`. It's name suggests that it is the module in question. Now, from your Arch image, can you run the following command, reboot and then report back with the results?

```bash
cat <<EOF > /etc/modules-load.d/wlan_ecr6600u_usb
# Load the bundled USB WiFi dongle's module at boot
wlan_ecr6600u_usb
EOF
```

After the reboot, can you please provide the output for the following commands?

```bash
ip link show

nmcli device status
```

**Comment by ankurkotwal** [comment]
> I assume that the output of `lsmod` was provided form the Debian image (I don't recall mentioning this earlier, apologies). In the provided output, I see the module `wlan_ecr6600u_usb`. It's name suggests that it is the module in question. Now, from your Arch image, can you run the following command, reboot and then report back with the results?

Yes I did it on Debian

> 
> ```shell
> cat <<EOF > /etc/modules-load.d/wlan_ecr6600u_usb
> # Load the bundled USB WiFi dongle's module at boot
> wlan_ecr6600u_usb
> EOF
> ```
> 
> After the reboot, can you please provide the output for the following commands?
> 
> ```shell
> ip link show
> 
> nmcli device status
> ```

That didn't work.
```
[root@vasudev ~]# ip link show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen0
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: end0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT group defaul0
    link/ether 6c:cf:39:00:28:27 brd ff:ff:ff:ff:ff:ff
3: end1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT group defaul0
    link/ether 6c:cf:39:00:28:28 brd ff:ff:ff:ff:ff:ff


[root@vasudev ~]# nmcli device status
DEVICE  TYPE      STATE        CONNECTION 
end0    ethernet  unavailable  --         
end1    ethernet  unavailable  --         
lo      loopback  unmanaged    --         

[root@vasudev ~]# lsmod
Module                  Size  Used by
wlan_ecr6600u_usb     339968  0
starfive_mailbox_test    16384  0
starfive_mailbox       16384  2

[root@vasudev ~]# journalctl --boot | grep wlan
Feb 18 06:12:14 vasudev kernel: wlan_ecr6600u_usb-ECR6600U_V1.1.0B04P05-build:200
23-03-01 12:50:34 CST
```

On Debian, I was able to do the following:
```
ankur@starfive /s/c/net> sudo lshw -C network
  *-network:0               
       description: Ethernet interface
       physical id: 5

       configuration: autonegotiation=on broadcast=yes driver=st_gmac driverversion=Jan_2016 link=n
  *-network:1
       description: Ethernet interface
       physical id: 6
       logical name: end1
       serial: 6c:cf:39:00:28:28
       size: 10Mbit/s
       capacity: 100Mbit/s
       capabilities: ethernet physical tp mii 10bt 10bt-fd 100bt 100bt-fd autonegotiation
       configuration: autonegotiation=on broadcast=yes driver=st_gmac driverversion=Jan_2016 duplex
  *-network:2 DISABLED
       description: Wireless interface
       physical id: 7
       bus info: usb@1:1.1
       logical name: wlx2c0547a11963
       serial: 9a:03:db:1b:d2:cf
       capabilities: ethernet physical wireless
       configuration: broadcast=yes driver=eswin_usb driverversion=5.15.0-starfive multicast=yes wi
  *-network:3
       description: Wireless interface
       physical id: 8
       bus info: usb@1:1.1
       logical name: wlx2c0547a11962
       serial: 2c:05:47:a1:19:62
       capabilities: ethernet physical wireless
       configuration: broadcast=yes driver=eswin_usb driverversion=5.15.0-starfive ip=192.168.86.53
```

Based on this, it looks like we need module eswin_usb.

**Comment by thefossguy** [comment]
Hmm, looks like this needs more effort. I'll plop in the Vendor image in my second baord and get back to you soon! :)

**Comment by thefossguy** [comment]
Hi @ankurkotwal, can you put [this firmware binary](https://github.com/eswincomputing/eswin_6600u/raw/master/firmware/ECR6600S_transport.bin) and place it inside `/lib/firmware`, reboot and provide the output of `dmesg | grep eswin`? Thanks :)

**Comment by ankurkotwal** [comment]
ip and nmcli still don't show the interface.

```
[root@vasudev ~]# dmesg | grep eswin
[   12.936313] [ecrnx] eswin_usb_init entry !!
[   12.940586] [ecrnx] eswin_usb_probe entry, reset slave !!
[   13.157666] [ecrnx] eswin_usb_probe entry, func: 0, g_usb: 0000000000000000 !!
[   13.163281] [ecrnx] eswin_tx_comp_thread entry
[   13.175000] [ecrnx] eswin_usb_probe kthread_run success.
[   13.175004] [ecrnx] eswin_usb_probe exit!!
[   13.175012] [ecrnx] eswin_rx_comp_thread entry
[   13.180430] [ecrnx] eswin_usb_probe entry, func: 1, g_usb: (____ptrval____) !!
[   13.188974] usbcore: registered new interface driver eswin_usb
[   13.202061] [ecrnx] eswin_usb_init exit !!
[   13.209977] [ecrnx] eswin_fw_file_chech, Checking firmware... (ECR6600U_transport.bin)
[   13.229119] eswin_usb 1-1.1:1.0: Direct firmware load for ECR6600U_transport.bin failed with error -2
[   13.238371] [ecrnx] eswin_fw_file_chech, error status = -2
```

**Comment by thefossguy** [comment]
Whoops, apologies, I linked to the wrong firmware binary. The link I provided was for ECR6600S. Taking a closer look at the loaded module in the vendor image, it is ECR6600**U**. (Don't remove the "wrong" firmware bin though.)

[Here is the [hopefully] correct link](https://github.com/eswincomputing/eswin_6600u/raw/master/firmware/ECR6600U_transport.bin). 😅

**Comment by ankurkotwal** [comment]
Thanks so much, we have progress! Unfortunately, it's still not quite going yet. The wifi scan results in an empty list and I can't manually connect to my access point

```
[root@vasudev ~]# ip link
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: end0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT group default qlen 1000
    link/ether 6c:cf:39:00:28:27 brd ff:ff:ff:ff:ff:ff
3: end1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT group default qlen 1000
    link/ether 6c:cf:39:00:28:28 brd ff:ff:ff:ff:ff:ff
4: wlp1s0u1u3: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN mode DEFAULT group default qlen 1000
    link/ether 1e:9c:6c:54:92:33 brd ff:ff:ff:ff:ff:ff permaddr 2c:05:47:a1:19:63
```

```
[root@vasudev ~]# nmcli dev status
DEVICE      TYPE      STATE        CONNECTION 
end0        ethernet  unavailable  --         
end1        ethernet  unavailable  --         
wlp1s0u1u3  wifi      unavailable  --         
lo          loopback  unmanaged    --         
```

```
[root@vasudev ~]# nmcli radio wifi
enabled
```

No access points found.
```
[root@vasudev ~]# nmcli dev wifi list                                                                                                                                                                      
IN-USE  BSSID  SSID  MODE  CHAN  RATE  SIGNAL  BARS  SECURITY 
[root@vasudev ~]# 
 ```

dmesg output
```[   12.960759] [ecrnx] eswin_usb_init entry !!
[   12.965009] [ecrnx] eswin_usb_probe entry, reset slave !!
[   13.177693] [ecrnx] eswin_usb_probe entry, func: 0, g_usb: 0000000000000000 !!
!
[   13.183315] [ecrnx] eswin_tx_comp_thread entry
[   13.183352] [ecrnx] eswin_usb_probe kthread_run success.
[   13.190593] [ecrnx] eswin_rx_comp_thread entry
[   13.195014] [ecrnx] eswin_usb_probe exit!!
[   13.195109] [ecrnx] eswin_usb_probe entry, func: 1, g_usb: (____ptrval____) !!
!
[   13.222073] [ecrnx] eswin_usb_init exit !!
[   13.227689] [ecrnx] eswin_fw_file_chech, Checking firmware... (ECR6600U_transs
port.bin)
[   13.254077] [ecrnx] eswin_fw_file_chech, request fw OK and size is 149980
[   13.261007] [ecrnx] fw_check_head,firmware build time: 2023-02-23 06:07:20
[   13.267944] [ecrnx] fw_check_head,firmware information: (ECR6600U_ECR6600U_V11
.1.0B05P06)
[   13.276157] [ecrnx] dl-fw >> sync, ret: 0
[   13.280577] [ecrnx] dl-fw >> sync, ret: 0
[   13.285169] [ecrnx] firmware downloading 1%  
[   13.289690] [ecrnx] firmware downloading 2% 
[   13.294203] [ecrnx] firmware downloading 3% 
[   13.298749] [ecrnx] firmware downloading 4%  
...
[   13.746498] [ecrnx] firmware downloading 98%
[   13.752090] [ecrnx] firmware downloading 99%
[   13.756693] [ecrnx] firmware downloading 100%
[   13.761289] [ecrnx] fimeware download successfully!
[   14.812025] [ecrnx] MAC Address is:00:11:22:33:44:5f
[   14.816998] [ecrnx] host driver log level is:3
[   14.821551] [ecrnx] firmware log level is:2
[   14.825826] [ecrnx] firmware log level type:0 (print to chip's uart)
[   14.863874] [ecrnx] ecrnx_fw_log_level_set: fstype:1, level:2, dir:0
[   14.871309] [ecrnx] info_len:12
[   21.047774] [ecrnx] cmd timed-out queue_sz:1
[   21.052108] [ecrnx] tkn[5]  flags:001a  result: -4  cmd:9730-ME_CHAN_CONFIG_RR
EQ       - reqcfm(9731-ME_CHAN_CONFIG_CFM)
[   21.186308] [ecrnx] \x013drivers/net/wireless/eswin/ipc_host.c:711:ASSERT_ERRR
(!env->msga2e_hostid)
[   27.447686] [ecrnx] cmd timed-out queue_sz:1
[   27.451973] [ecrnx] tkn[6]  flags:001a  result: -4  cmd:8194-MM_START_REQ     
         - reqcfm(8195-MM_START_CFM)
[   27.464349] [ecrnx] \x013drivers/net/wireless/eswin/ipc_host.c:711:ASSERT_ERRR
(!env->msga2e_hostid)
[   33.847754] [ecrnx] cmd timed-out queue_sz:1
[   33.852087] [ecrnx] tkn[7]  flags:001a  result: -4  cmd:8194-MM_START_REQ     
         - reqcfm(8[   33.862711] [ecrnx] \x013drivers/net/wireless/eswin/ipc_host.c:711:ASSERT_ERRR
(!env->msga2e_hostid)
[   40.247758] [ecrnx] cmd timed-out queue_sz:1
[   40.252128] [ecrnx] tkn[8]  flags:001a  result: -4  cmd:8194-MM_START_REQ     
         - reqcfm(8195-MM_START_CFM)
[   40.277915] [ecrnx] \x013drivers/net/wireless/eswin/ipc_host.c:711:ASSERT_ERRR
(!env->msga2e_hostid)
[   46.647768] [ecrnx] cmd timed-out queue_sz:1
[   46.652119] [ecrnx] tkn[9]  flags:001a  result: -4  cmd:8194-MM_START_REQ     
         - reqcfm(8195-MM_START_CFM)
[   46.672606] [ecrnx] \x013drivers/net/wireless/eswin/ipc_host.c:711:ASSERT_ERRR
(!env->msga2e_hostid)
[   53.057706] [ecrnx] cmd timed-out queue_sz:1
[   53.061999] [ecrnx] tkn[10]  flags:001a  result: -4  cmd:8194-MM_START_REQ    
          - reqcfm(8195-MM_START_CFM)
[   53.074679] [ecrnx] \x013drivers/net/wireless/eswin/ipc_host.c:711:ASSERT_ERRR
(!env->msga2e_hostid)
195-MM_START_CFM)
```


Thanks so much for looking into this. Please let me know what you'd like me to look into next.

**Comment by thefossguy** [comment]
Closing this as this is now addressed in thefossguy/linux-starfive-visionfive2@debc93f73acf7abeac05a822a398835a4b730ccd.

**Comment by thefossguy** [comment]
Fixed in thefossguy/archlinux-visionfive2@950e7902e3e2738feda95f531f561442276b2a28
(**not yet tested**)
